akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazySinkSpec.scala

```diff
@@ -0,0 +1,135 @@
+      (1 to 10).foreach(i ⇒ {
+        sourceSub.sendNext(i)
+        probe.expectNext(i)
+      })
+      sourceSub.sendComplete()
+    }
+
+    "complete when there was no elements in stream" in assertAllStagesStopped {
+      val futureProbe = Source.empty.runWith(Sink.lazySink[Int, Future[Int]](_ ⇒ Future.successful(Sink.fold[Int, Int](0)(_ + _)), () ⇒ Future.successful(0)))
+      val futureResult = Await.result(futureProbe, 300.millis)
+      Await.result(futureResult, 300.millis) should ===(0)
+    }
+
+    "complete normally when upstream is completed" in assertAllStagesStopped {
+      val futureProbe = Source.single(1).runWith(Sink.lazySink[Int, Probe[Int]](_ ⇒ Future.successful(TestSink.probe[Int]), () ⇒ ???))
+      val futureResult = Await.result(futureProbe, 300.millis)
+      futureResult.request(1)
+        .expectNext(1)
+        .expectComplete()
+    }
+
+    "failed gracefully when sink factory method failed" in assertAllStagesStopped {
+      val sourceProbe = TestPublisher.manualProbe[Int]()
+      val futureProbe = Source.fromPublisher(sourceProbe).runWith(Sink.lazySink[Int, Probe[Int]](_ ⇒ throw ex, () ⇒ ???))
+
+      val sourceSub = sourceProbe.expectSubscription()
+      sourceSub.expectRequest(1)
+      sourceSub.sendNext(0)
+      sourceSub.expectCancellation()
+      a[RuntimeException] shouldBe thrownBy { Await.result(futureProbe, 300.millis) }
+    }
+
+    "failed gracefully when upstream failed" in assertAllStagesStopped {
+      val sourceProbe = TestPublisher.manualProbe[Int]()
+      val futureProbe = Source.fromPublisher(sourceProbe).runWith(Sink.lazySink[Int, Probe[Int]](_ ⇒ Future.successful(TestSink.probe[Int]), () ⇒ ???))
+
+      val sourceSub = sourceProbe.expectSubscription()
+      sourceSub.expectRequest(1)
+      sourceSub.sendNext(0)
+      val probe = Await.result(futureProbe, 300.millis)
+      probe.request(1)
+        .expectNext(0)
+      sourceSub.sendError(ex)
+      probe.expectError(ex)
+    }
+
+    "cancel upstream when internal sink is cancelled" in assertAllStagesStopped {
+      val sourceProbe = TestPublisher.manualProbe[Int]()
+      val futureProbe = Source.fromPublisher(sourceProbe).runWith(Sink.lazySink[Int, Probe[Int]](_ ⇒ Future.successful(TestSink.probe[Int]), () ⇒ ???))
+      val sourceSub = sourceProbe.expectSubscription()
+      sourceSub.expectRequest(1)
+      sourceSub.sendNext(0)
+      sourceSub.expectRequest(1)
+      val probe = Await.result(futureProbe, 300.millis)
+      probe.request(1)
+        .expectNext(0)
+      probe.cancel()
+      sourceSub.expectCancellation()
+    }
+
+    "continue if supervision is resume" in assertAllStagesStopped {
+      val sourceProbe = TestPublisher.manualProbe[Int]()
+      val futureProbe = Source.fromPublisher(sourceProbe).runWith(Sink.lazySink[Int, Probe[Int]](a ⇒
+        if (a == 0) throw ex else Future.successful(TestSink.probe[Int]), () ⇒ ???)
+        .withAttributes(supervisionStrategy(resumingDecider)))
```

Comment:
Hm in case Sink is depending on element we passing to factory method - than yes.
But perhaps this case should be done not through failing and resuming but though Option return type for factory function. What do you think?